### PR TITLE
DAITDS-231 - Develop TSIG support in Go SDK

### PR DESCRIPTION
**Pull request by ThiagoYU**

### CODE REVIEW

File: rest/model/dns/tsig_key.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package dns
+
+// Zone wraps an NS1 /zone resource
+type Tsig_key struct {
```
Comment (Zach-Johnson) [inline rest/model/dns/tsig_key.go]: No underscores in var/struct names in Go - lets make this `TSIGKey`

File: rest/model/dns/tsig_key.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package dns
+
+// Zone wraps an NS1 /zone resource
```
Comment (Zach-Johnson) [inline rest/model/dns/tsig_key.go]: Fix comment

File: rest/model/dns/tsig_key.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+type Tsig_key struct {
+	Name      string `json:"name,omitempty"`
+	Algorithm string `json:"algorithm,omitempty"`
+	Secret    string `json:"secret,omitempty"`
+}
+
+// NewTSIGKey takes a name, algorithm and secret and creates a new zone.
```
Comment (Zach-Johnson) [inline rest/model/dns/tsig_key.go]: Fix comment

File: rest/tsig_key.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+	req, err := s.client.NewRequest("GET", path, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var tk dns.Tsig_key
+	var resp *http.Response
+	resp, err = s.client.Do(req, &tk)
+
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: remove extra line

File: rest/tsig_key.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	if err != nil {
+		switch errType := err.(type) {
+		case *Error:
+			if errType.Message == fmt.Sprintf("TSIG %s. was not found", name) {
+				return nil, resp, ErrTsigKeyMissing
+			}
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: can we check for a 404 status code on this instead?

File: rest/tsig_key.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	// Update TSIG key fields with data from api(ensure consistent)
+	resp, err := s.client.Do(req, &tk)
+	if err != nil {
+		switch errType := err.(type) {
+		case *Error:
+			if errType.Message == fmt.Sprintf("TSIG %s. already exists", tk.Name) {
+				return resp, ErrTsigKeyExists
+			}
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: Is there a status code for this error that we could check for instead?

File: rest/tsig_key.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	// Update TSIG key fields with data from api(ensure consistent)
+	resp, err := s.client.Do(req, &tk)
+	if err != nil {
+		switch errType := err.(type) {
+		case *Error:
+			if errType.Message == fmt.Sprintf("TSIG %s. was not found", tk.Name) {
+				return resp, ErrTsigKeyMissing
+			}
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: can we check for 404 status code on this?

File: rest/tsig_key.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+		return nil, err
+	}
+
+	resp, err := s.client.Do(req, nil)
+	if err != nil {
+		switch errType := err.(type) {
+		case *Error:
+			if errType.Message == fmt.Sprintf("TSIG %s. was not found", name) {
+				return resp, ErrTsigKeyMissing
+			}
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: can we check for 404 status code on this?

**Review by Zach-Johnson — CHANGES_REQUESTED** [review]
Looks good, just a few clean up things.

File: rest/model/dns/tsig_jey_test.go [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package dns
```
Comment (dmavrommatis) [inline rest/model/dns/tsig_jey_test.go]: name of the file has a typo

**Review by dmavrommatis — CHANGES_REQUESTED** [review]
can you please run staticcheck for your changes to see lint issues?

File: rest/tsig_key.go [hunk 10]
```diff
@@ -136,6 +138,13 @@ func (s *TsigService) Delete(name string) (*http.Response, error) {
 }
 
 var (
+	// Status Codes used
+	BadRequestCode          = 400
+	NotFoundStatusCode      = 404
+	AlreadyExistsStatusCode = 409
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: Please use the existing constants in stdlib, they are in the http package. e.g. `http.StatusNotFound`

File: rest/tsig_key.go [hunk 11]
```diff
@@ -136,6 +138,13 @@ func (s *TsigService) Delete(name string) (*http.Response, error) {
...
+
+	// ErrTsigBadRequest
+	ErrTsigBadRequest = errors.New("bad request error, expected in S1 (18/10/2021)")
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: Let's just return the standard error for a bad request, we don't need to expose this.

File: rest/model/dns/tsig_key.go [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+// TSIGKey wraps an NS1 /tsig resource
+type TSIGKey struct {
+	Name      string `json:"name,omitempty"`
+	Algorithm string `json:"algorithm,omitempty"`
+	Secret    string `json:"secret,omitempty"`
+}
+
+// NewTSIGKey takes a name, algorithm and secret and creates a new TSIG key.
+func NewTsigKey(name string, algorithm string, secret string) *TSIGKey {
+	tsig_key := TSIGKey{
```
Comment (Zach-Johnson) [inline rest/model/dns/tsig_key.go]: No underscores in var names. Please run a go linter on this code, it should point out stuff like this.

File: rest/tsig_key.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+				return resp, ErrTsigBadRequest
+			}
+		}
+		return resp, err
+	}
+
+	return resp, nil
+}
+
+// Update takes a *Tsug_key and modifies basic details of a TSIG key.
```
Comment (Zach-Johnson) [inline rest/tsig_key.go]: typo `Tsug_key`

**Review by Zach-Johnson — CHANGES_REQUESTED** [review]

File: rest/tsig_key.go (hunk 11 above)
Comment (Zach-Johnson) [reply rest/tsig_key.go]: @ThiagoYU this comment is unresolved still

**Review by Zach-Johnson — CHANGES_REQUESTED** [review]

File: rest/tsig_key.go (hunk 11 above)
Comment (ThiagoYU) [reply rest/tsig_key.go]: @Zach-Johnson Sorry for this... Just fixed

**Review by pcatalini — APPROVED** [review]
LGTM

**Review by Zach-Johnson — APPROVED** [review]

**Pull request merged**